#5: Added thirdparty rules E1, E3, and E4

vlad961 opened the pull request:
#5 
Implemented Thirdparty rules E1, E2 and E3 with Archunit.

vlad961 committed:
Added E1,E3 and E4

vlad961 committed:
Refactored error output to see the violated rule.

vlad961 committed:
Merge branch 'devonfw-sample:master' into add-thirdparty-rules

hohwille committed:
Merge branch 'master' into add-thirdparty-rules

hohwille commented on src/test/java/com/devonfw/sample/archunit/ThirdPartyRulesE1TransactionalTest.java:
```diff
@@ -0,0 +1,57 @@
+@AnalyzeClasses(packages = "com.devonfw.sample.archunit", importOptions = ImportOption.DoNotIncludeTests.class)
+public class ThirdPartyRulesE1TransactionalTest {
+
+    private static boolean apiScopedClassIsUsingTransactional(JavaClass item, String targetPackageFullName) {
```
nice that you introduce methods with speaking names to increase readability of the code.
To align with the other method and Java best practices, I would consider renaming to `isApiScopedClassUsingTransactional`.

vlad961 replied:
Refactored the name to align with Java best practices.

vlad961 replied:
Deleted all scope related tests and postponed a possible reimplementation after a  [(`DevonPackageImpl`)](https://github.com/devonfw/sonar-devon4j-plugin/blob/master/plugin/src/main/java/com/devonfw/ide/sonarqube/common/api/config/DevonPackageImpl.java) like class is present in our project.

hohwille replied:
Thanks.

hohwille commented on src/test/java/com/devonfw/sample/archunit/ThirdPartyRulesE1TransactionalTest.java:
```diff
@@ -0,0 +1,57 @@
+public class ThirdPartyRulesE1TransactionalTest {
+
+    private static boolean apiScopedClassIsUsingTransactional(JavaClass item, String targetPackageFullName) {
+        if(item.getFullName().contains("api") && targetPackageFullName.equals("javax.transaction.Transactional")) {
+                return true;
+        }
```
Actually this is from the original sonar plugin rule.
We have decided to go for the modern and simplified packaging and project structure:
https://github.com/devonfw/devon4j/blob/master/documentation/guide-structure-modern.asciidoc
whereas the sonar plugin rules were based on the legacy packaging and project structure:
https://github.com/devonfw/devon4j/blob/master/documentation/guide-structure-classic.asciidoc

so before we had e.g. and interface `...logic.api.UcFindTask` and its implementation `...logic.impl.UcFindTaskImpl` whereas we now simplified and only have an implementation `...logic.UcFindTask` and its public methods are also its API. You will not find any `api` package in our sample app hence we should remove all scope related aspects here as they will not make sense and rather cause confusion.

vlad961 replied:
Deleted all scope related tests and postponed a possible reimplementation after a  [(`DevonPackageImpl`)](https://github.com/devonfw/sonar-devon4j-plugin/blob/master/plugin/src/main/java/com/devonfw/ide/sonarqube/common/api/config/DevonPackageImpl.java) like class is present in our project.

hohwille replied:
Fine for now. Can later be reworked with `StructuredPackage`.

hohwille commented on src/test/java/com/devonfw/sample/archunit/ThirdPartyRulesE1TransactionalTest.java:
```diff
@@ -0,0 +1,57 @@
+        return false;
+    }
+
+    static ArchCondition<JavaClass> miuse_springframework_transactional_annotation = new ArchCondition<JavaClass> ("misuse @Transactional (Rule-E1)") {
```
typo `miuse` should be `misuse`. Also should be `final`. 

FYI: For conventions we should align in the team as everybody creates its own style but we can resolve this with story #19

vlad961 replied:
Refactored `miuse` to `misuse`

hohwille replied:
Thanks.

hohwille commented on src/test/java/com/devonfw/sample/archunit/ThirdPartyRulesE1TransactionalTest.java:
```diff
@@ -0,0 +1,57 @@
+        @Override
+        public void check(JavaClass item, ConditionEvents events) {
+                for(Dependency access: item.getDirectDependenciesFromSelf()) {
+                        String targetPackageFullName = access.getTargetClass().getFullName();
```
The `Package` in the variable name is IMHO misleading as it is a qualified name of a type / class and not just a package name.

vlad961 replied:
Indeed, that's a poor name choice. I changed the variable name `targetPackageFullName` to `targetFullName`. 
In addition, I changed the first parameter name of the _check_ method from `item` to `sourceClass,` which leads to better code readability and understanding.
Also, I changed the Dependency name in Row:35 from `access` to `dependency,` since access might also be misleading. One might think there's actual access towards a forbidden class, whereas even declaring a field of an undesired class type will resolve into an error message.

hohwille replied:
OK. Thanks. Great.

hohwille commented on src/test/java/com/devonfw/sample/archunit/ThirdPartyRulesE3JpaCheckTest.java:
```diff
@@ -0,0 +1,50 @@
+            if(item.getFullName().contains("dataaccess")) {
+                return true;
+            }
+            if(item.getFullName().contains("common") && item.getSimpleName().contains("Embeddable")) {
```
it would be better to reuse our regular expression and have some reusable code to check layer and other aspects of the package. Especially this rule would also apply in case someone has an entity `com.customer.app.commoncomponent.dataaccess.MyCommonEntity` as the qualified name contains `common`.

vlad961 replied:
Concerning #19, we should discuss with the team how we want to support the reusing of given regular expressions.

For the moment, I have added two `public static final` patterns for the common and dataaccess layers into `PackageRuleTest.java`. I utilize these patterns to check whether the source class lies inside the common or dataaccess layer instead of that faulty implementation.

hohwille replied:
same here. We can merge now, but has to be reworked anyhow. `StructuredPackage` now available.

hohwille commented on src/test/java/com/devonfw/sample/archunit/ThirdPartyRulesE3JpaCheckTest.java:
```diff
@@ -0,0 +1,50 @@
+
+    private static boolean isUsingJavaxPersistenceDataAccessOrEmbeddablesInCommon(JavaClass item, String targetPackageFullName) {
+        if(targetPackageFullName.startsWith("javax.persistence")) {
+            if(item.getFullName().contains("dataaccess")) {
```
same here even though `dataaccess` is quite specific and it would be rather odd to hit this by accident (e.g. `com.customer.dataaccessmanagement.logic.MyCustomEntity`).

vlad961 replied:
Used a hotfix. See above comment.

hohwille replied:
OK. StructuredPackage now available. Left for rework after merge.

hohwille commented on src/test/java/com/devonfw/sample/archunit/ThirdPartyRulesE4HibernateCheckTest.java:
```diff
@@ -0,0 +1,110 @@
+    }
+
+    private static boolean isNotImplementingHibernateEnversInImplScope(JavaClass item, ConditionEvents events, String targetPackageName, String targetPackageFullName, String targetClassDescription, String targetSimpleName) {
+        if(targetPackageName.startsWith(ORG_HIBERNATE_ENVERS) && !item.getPackageName().contains("impl")
```
once again a `scope` check that we need to eliminate.

vlad961 replied:
Deleted all scope related tests and postponed a possible reimplementation after a  [(`DevonPackageImpl`)](https://github.com/devonfw/sonar-devon4j-plugin/blob/master/plugin/src/main/java/com/devonfw/ide/sonarqube/common/api/config/DevonPackageImpl.java) like class is present in our project.

hohwille replied:
Thanks. Done.

hohwille reviewed (changes requested):
@vlad961 thanks for your PR. This was actually quite a complex story to implement. You have done well to figure out the magic tricks with `item.getDirectDependenciesFromSelf()`, etc. Great job 👍 

Obviously when it is complex I do have review comments for rework. Can you have a look.
I would also suggest that we have a small meeting together end of this week or start of next week to discuss on this. The adaptation to get rid of the scope might not be trivial without further assistance.

BTW: Can you check your IDE setup? With devonfw-ide and Eclipse you should have save actions enabled that give you automatic code-formatting when you save. Are you using IntelliJ or VSCode instead? Fine as well but then also ensure you format the source-code and configure your formatter using devonfw-code-style (2 spaces, opening curly braces on the same line, etc.).

vlad961 committed:
Update ThirdPartyRulesE1TransactionalTest.java

Refactored the class with respect to @hohwille review annotations.

vlad961 committed:
Refactoring

Concerning @hohwille review annotation, I adjusted the E3 Rule to correctly check for layers.
I added two public static final patterns for the common and dataaccess layers to use instead of the faulty contains() check which would trigger false positive errors.

vlad961 committed:
Update ThirdPartyRulesE3JpaCheckTest.java

changed misuse_jpa ArchCondition to final and refactored parameter name for better understandability

vlad961 committed:
Update ThirdPartyRulesE4HibernateCheckTest.java

Fixed the faulty layer check condition of the isUsingHibernateOutsideOfDataaccessLayer method.
Added the condition in row 97. (Was erroneously placed inside isUsingHibernateOutsideOfDataaccessLayer method)
Refactored class via automatic code-formatting of devonfw-ide.
Deleted unused method parameters.

vlad961 committed:
Removed unused common layer pattern.

vlad961 committed:
Deleted the scope related:
Transactional and Hibernate tests.
Postponed them until a concrete ArchitecturePackage impl is present.

vlad961 committed:
Set ArchCondition to final.

hohwille committed:
Merge branch 'master' into add-thirdparty-rules

hohwille reviewed (approved):
We do have some homework left to rework with #19 and also with the scope that can now be addressed with `PackageStructure`. Check the review comments to complete #5.

However, I will merge to resolve blockages with too many open PRs.
